firefox-ios/Client/Frontend/Library/Bookmarks/Utitlity/FolderHierarchyFetcher.swift

```diff
@@ -35,22 +35,23 @@ struct DefaultFolderHierarchyFetcher: FolderHierarchyFetcher, BookmarksRefactorF
 
     func fetchFolders() async -> [Folder] {
         let numDesktopBookmarks = await countDesktopBookmarks()
-        return await withCheckedContinuation { continuation in
-            profile.places.getBookmarksTree(rootGUID: rootFolderGUID,
-                                            recursive: true).uponQueue(.main) { data in
-                var folders = [Folder]()
-                defer {
-                    continuation.resume(returning: folders)
-                }
-                guard let rootFolder = data.successValue as? BookmarkFolderData else { return }
-                let hasDesktopBookmarks = (numDesktopBookmarks ?? 0) > 0
-
-                let childrenFolders = rootFolder.children?.compactMap {
-                    return $0 as? BookmarkFolderData
-                }
+        return await withUnsafeContinuation { continuation in
+            profile.places.getBookmarksTree(rootGUID: rootFolderGUID, recursive: true) { result in
+                switch result {
+                case .success(let data):
+                    guard let rootFolder = data as? BookmarkFolderData else { return }
+                    let hasDesktopBookmarks = (numDesktopBookmarks ?? 0) > 0
 
-                for folder in childrenFolders ?? [] {
-                    recursiveAddSubFolders(folder, folders: &folders, hasDesktopBookmarks: hasDesktopBookmarks)
+                    let childrenFolders = rootFolder.children?.compactMap {
+                        return $0 as? BookmarkFolderData
```

Comment:
We also don't call the continuation here now either

Reply:
We call `resume` on both `Result` cases.

Reply:
Oh I see, on the `guard` statement you're saying.

Reply:
Oops yeah you're right that this one is in a compactMap, but yeah the guard one is a problem!

Reply:
I will put `defer` as it was before then.

Reply:
Or just call the continuation in the guard would be reasonable, whichever you prefer!

If you use a defer, can you please check that the continuation doesn't get called twice with a breakpoint if possible? Since the `unsafe` version won't tell us if we're misusing the continuation, so we should be careful. 🙏  I'm worried the continuation resuming with `folders` below will _also_ be triggered in the good case, as well as the `defer`, calling `resume` twice...